### ISSUE TITLE
CA2000 false positive

### ISSUE THREAD

**Issue by avivanoff** [dotnet/roslyn-analyzers#2936]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.6 (Latest)

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope)

#### Repro steps

We have some false positives CA2000 (started happening after upgrade from 2.9.4 to 2.9.6). Unfortunately, I am not able to create a simple console app to reproduce the issue. Are there any logs/diagnostics/etc. I can provide to you to help diagnosing the issue?

**Comment by avivanoff** [comment]
Just upgraded to 2.9.7, still false positives CA2000. What can I do to diagnose and fix the issues?

**Comment by avivanoff** [comment]
One interesting observation that may help with diagnosing the issue. If I change
```
using (var msg = new MailMessage())
{
    ... a lot of complex async code
}
```
to
```
using var msg = new MailMessage();
... a lot of complex async code
```
then CA2000 does not happen anymore.

**Comment by avivanoff** [comment]
@mavasani, do you have any comments? This issue prevents from upgrading to the latest version of analyzers.

**Comment by mavasani** [comment]
We explicitly bail out on new using variable declaration style added in C# 8 because there was no CFG or IOperation for it until a few weeks ago. So, https://github.com/dotnet/roslyn-analyzers/issues/2936#issuecomment-550937680 is a red herring.

We would still need you to provide some kind of a repro for your false positive to investigate it.

**Comment by avivanoff** [comment]
@mavasani, as I mentioned in the original post, I spent a lot of time trying to create an app to reproduce the issue. Is there any other way? Logs? Live Share?

**Comment by avivanoff** [comment]
@mavasani, I was finally able to create a repro:
```
namespace ConsoleApp3
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Net.Mail;
    using System.Threading.Tasks;

    internal static class Program
    {
        #region Private implementation

        private static Attachment CreateAttachment(String filePath, String attachmentName)
        {
            var attachment = new Attachment(filePath);
            try
            {
                attachment.Name = Path.ChangeExtension(attachmentName, Path.GetExtension(filePath));

                return attachment;
            }
            catch (Exception)
            {
                attachment.Dispose();

                throw;
            }
        }

        private static async Task Main()
        {
            var recipients = new[] { "user1@me.com", "user2@me.com" };
            var attachments = new[] { "att1", "att2" };
            foreach (var recipient in recipients)
            {
                using (var message = new MailMessage())
                {
                    message.Subject = "Subj";
                    await Program.SendAsync(message, attachments, recipient)
                                 .ConfigureAwait(false);
                }
            }
        }

        private static async Task SendAsync(MailMessage message, IReadOnlyCollection<String> attachments, String recipient)
        {
            try
            {
                message.To.Add(recipient);
                foreach (var attachment in attachments)
                {
                    if (!String.IsNullOrEmpty(attachment))
                    {
                        message.Attachments.Add(Program.CreateAttachment(attachment, attachment));
                    }
                }
                using (var client = new SmtpClient())
                {
                    await client.SendMailAsync(message)
                                .ConfigureAwait(false);
                }
            }
            catch (SmtpException ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        #endregion
    }
}
```

CA2000 is triggered on line **using (var message = new MailMessage())**.

**Comment by avivanoff** [comment]
@mavasani, have you been able to reproduce the issue?

**Comment by mavasani** [comment]
Thanks for the repro @avivanoff. I haven't been able to work on a fix here as have been busy with other Roslyn work items. I hope to get to the bugs in this repo in a week or two.

**Comment by mavasani** [comment]
@avivanoff It seems to be a bug in interprocedural analysis. Can you please try to turn off interprocedural analysis: https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#interprocedural-analysis-kind

Also for reference: https://github.com/dotnet/roslyn-analyzers/blob/master/docs/Analyzer%20Configuration.md#enabling-editorconfig-based-configuration-for-a-project

**Comment by mavasani** [comment]
Unit test for repro:

```csharp
[Fact]
        public void UsingWithInterproceduralThrowExpression_NoDiagnostic()
        {
            VerifyCSharp(@"
using System;

internal class A : IDisposable
{
    public void Dispose()
    {
    }
}

internal class Test
{
    private static A CreateNew()
    {
        try
        {
            return new A();
        }
        catch (Exception)
        {
            throw;
        }
    }

    private static void M(string[] array)
    {
        foreach (var _ in array)
        {
            using (var a = new A())
            {
                using (var x = CreateNew())
                {
                }
            }
        }
    }
}
");
        }
```

**Comment by RoelofSpijker** [comment]
I've ran into a similar problem just now and after some searching came upon this issue. My case looks _slightly_ different, but I believe it's the same underlying issue. In my case it appears to occur when a method that *might* result in an `IDisposable` is being called from within a `using` from within a `foreach`. A "minimal" working example looks like:

```C#
class Program
    {
        static void Main()
        {
            foreach (Stream s in MakeIterators(new string[] { "FileA", "FileB" }))
            {
                Console.WriteLine($"First byte of stream: {s.ReadByte()}");
            }

            Console.ReadKey();
        }

        private static IEnumerable<Stream> MakeIterators(string[] fileNames)
        {
            using (ArchiveWrapper aw = new ArchiveWrapper("Non-looped")) /* This does not cause CA2000 */
            {
                yield return aw.GetStream();
            }

            foreach (string fileName in fileNames)
            {
                using (ArchiveWrapper aw = new ArchiveWrapper(fileName)) /* The call to new DisposableLogger here causes CA2000 */
                {
                    yield return aw.GetStream();
                }
            }
        }

        class ArchiveReader : IDisposable
        {
            public List<Stream> Entries = new List<Stream>();
            private readonly string fileName;

            public ArchiveReader(string fileName)
            {
                this.fileName = fileName;
                Entries.Add(new MemoryStream(Encoding.UTF8.GetBytes(fileName)));
            }

            public void Dispose()
            {
                Console.WriteLine($"Disposing {fileName} reader");
            }
        }

        class ArchiveWrapper : IDisposable
        {
            private IDisposable streamWrapper;
            private readonly string fileName;

            public ArchiveWrapper(string fileName)
            {
                this.fileName = fileName;
            }

            public Stream GetStream()
            {
                ArchiveReader ar = new ArchiveReader(fileName);
                streamWrapper = ar;
                /* This potential exception is crucial */
                if (ar.Entries.Any())
                    return ar.Entries.First();
                else
                    throw new Exception();
            }

            public void Dispose()
            {
                Console.WriteLine($"Disposing {fileName} wrapper");
                streamWrapper.Dispose();
            }
        }
    }
```

I'm not 100% sure that it is the same issue as identified here, but it does look very similar. I'm seeing this on 2.9.6 as well as on 2.9.8. Setting `interprocedural_analysis_kind` to `none` does cause the warning to no longer appear.
